language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/semantictokens/SemanticTokensVisitor.java

```diff
@@ -0,0 +1,797 @@
+     *
+     * @param node      Current node
+     * @param startLine Start line position
+     */
+    private void processSymbols(Node node, LinePosition startLine) {
+        if (!semanticTokens.contains(new SemanticToken(startLine.line(), startLine.offset()))) {
+            Optional<SemanticModel> semanticModel;
+            try {
+                semanticModel = this.semanticTokensContext.currentSemanticModel();
+            } catch (Throwable e) {
+                return;
+            }
+            if (semanticModel.isEmpty()) {
+                return;
+            }
+            Optional<Symbol> symbol = semanticModel.get().symbol(node);
+            if (symbol.isPresent() && symbol.get().getLocation().isPresent()) {
+                LineRange symbolLineRange = symbol.get().getLocation().get().lineRange();
+                LinePosition linePosition = symbolLineRange.startLine();
+                SymbolKind kind = symbol.get().kind();
+                String nodeName = node.toString().trim();
+
+                int declarationType = -1, declarationModifiers = -1, referenceType = -1, referenceModifiers = -1;
+                switch (kind) {
+                    case CLASS:
+                        if (!nodeName.equals(SemanticTokensConstants.SELF)) {
+                            declarationType = TokenTypes.CLASS.getId();
+                            declarationModifiers = TokenTypeModifiers.DECLARATION.getId();
+                            referenceType = TokenTypes.CLASS.getId();
+                        }
+                        break;
+                    case CLASS_FIELD:
+                        declarationType = TokenTypes.PROPERTY.getId();
+                        declarationModifiers = TokenTypeModifiers.DECLARATION.getId();
+                        referenceType = TokenTypes.PROPERTY.getId();
+                        break;
+                    case CONSTANT:
+                        declarationType = TokenTypes.VARIABLE.getId();
+                        declarationModifiers = TokenTypeModifiers.DECLARATION.getId() |
+                                TokenTypeModifiers.READONLY.getId();
+                        referenceType = TokenTypes.VARIABLE.getId();
+                        referenceModifiers = TokenTypeModifiers.READONLY.getId();
+                        break;
+                    case VARIABLE:
+                        boolean isReadonly = false;
+                        if (symbol.get() instanceof VariableSymbol && ((VariableSymbol) symbol.get()).typeDescriptor()
+                                instanceof IntersectionTypeSymbol && ((IntersectionTypeSymbol) ((VariableSymbol) symbol
+                                .get()).typeDescriptor()).memberTypeDescriptors().stream().anyMatch(desc ->
+                                desc.typeKind() == TypeDescKind.READONLY)) {
+                            isReadonly = true;
+                        }
+                        if (!nodeName.equals(SemanticTokensConstants.SELF)) {
```

Comment:
If we strictly avoid the self keyword, shall we get this check to the top? So there will be no need of checking this within specific cases